lib/sms/sms_deliver.c

```diff
@@ -0,0 +1,663 @@
+			}
+			sprintf(encoded_number + hex_str_index + 1,
+				"%d", number);
+		}
+		hex_str_index += 2;
+	}
+	encoded_number[hex_str_index] = '\0';
+
+	LOG_DBG("Address-Value: %s", log_strdup(encoded_number));
+
+	memset(sms_header->originating_address.address_str, 0,
+		SMS_MAX_ADDRESS_LEN_CHARS + 1);
+	memcpy(sms_header->originating_address.address_str, encoded_number,
+		SMS_MAX_ADDRESS_LEN_CHARS + 1);
+
+	sms_header->originating_address.length =
+		DELIVER_DATA(parser)->field_oa.length;
+	sms_header->originating_address.type   =
+		DELIVER_DATA(parser)->field_oa.adr_type;
+
+	memcpy(sms_header->originating_address.address,
+	       DELIVER_DATA(parser)->field_oa.adr,
+	       SMS_MAX_ADDRESS_LEN_OCTETS);
+
+	sms_header->ud_len = DELIVER_DATA(parser)->field_udl -
+			     DELIVER_DATA(parser)->field_udhl;
+
+	if (DELIVER_DATA(parser)->field_udhl > 0) {
+		sms_header->udh_len = DELIVER_DATA(parser)->field_udhl - 1;
+		sms_header->udh = k_malloc(DELIVER_DATA(parser)->field_udhl);
+		if (sms_header->udh == NULL) {
+			LOG_ERR("Parsing failed due to no memory");
+			return -ENOMEM;
+		}
+		memcpy(sms_header->udh, DELIVER_DATA(parser)->field_udh, DELIVER_DATA(parser)->field_udhl);
+	} else {
+		sms_header->udh_len = 0;
+		sms_header->udh = NULL;
+	}
+
+	sms_header->app_port = DELIVER_DATA(parser)->field_udh_app_port;
+	sms_header->concatenated = DELIVER_DATA(parser)->field_udh_concatenated;
+
+	return 0;
+}
+
+const static struct parser_api sms_deliver_api = {
+	.data_size        = sms_deliver_get_data_size,
+	.get_parsers      = sms_deliver_get_parsers,
+	.get_decoder      = sms_deliver_get_decoder,
+	.get_parser_count = sms_deliver_get_parser_count,
+	.get_header       = sms_deliver_get_header,
+};
+
+void *sms_deliver_get_api(void)
+{
+	return (struct parser_api*)&sms_deliver_api;
+}
+
+/** 
+ * @brief Decode received SMS message, i.e., SMS-DELIVER message as specified
+ * in 3GPP TS 23.040 Section 9.2.2.1.
+ * 
+ * @param pdu SMS-DELIVER PDU.
+ * @param out SMS message decoded into a structure.
+ * 
+ * @retval -EINVAL Invalid parameter.
+ * @retval -ENOMEM No memory to register new observers.
+ * @return Zero on success, otherwise error code.
+ */
+int sms_deliver_pdu_parse(char *pdu, struct sms_deliver_header *out)
+{
+	struct  parser sms_deliver;
+	int     err=0;
+
+	__ASSERT(pdu != NULL, "Parameter 'pdu' cannot be NULL.");
+	__ASSERT(out != NULL, "Parameter 'out' cannot be NULL.");
+
+	memset(out, 0, sizeof(struct sms_deliver_header));
+
+	parser_create(&sms_deliver, sms_deliver_get_api());
+
+	err = parser_process_str(&sms_deliver, pdu);
+
+	if (err) {
+		LOG_ERR("Parsing error (%d) in decoding SMS-DELIVER message due to no memory", err);
+		return err;
+	}
+
+	parser_get_header(&sms_deliver, out);
+
+	out->ud = k_calloc(1, out->ud_len + 1);
+	if (out->ud == NULL) {
+		LOG_ERR("Unable to parse SMS-DELIVER message due to no memory");
+		return -ENOMEM;
+	}
+
+	out->data_len = parser_get_payload(&sms_deliver,
```

Comment:
This is still decoding payload also when the message is received.
If we decide to leave separate get_payload function for the client to call, do we require client also to reserve memory based on User-Data-Length?
Notice that the actual data can be smaller and is smaller in case of GSM 7bit encoding. And we cannot know the exact length before performing the actual decoding.

Reply:
This has changed drastically due to refactoring and is not valid as such anymore